File: integration_tests/typescript_client_tests/test/mock-adapter.ts
// Local copy of the interface — structurally compatible with the generated one.
export interface GraphLinkWsAdapter {
  connect(): Promise<void>;
  sendMessage(message: string): Promise<void>;
  get onMessageStream(): AsyncIterable<string>;
  close(): Promise<void>;
  connectionInitPayload(): Promise<Record<string, unknown> | null>;
  get onReconnect(): AsyncIterable<void>;
}

export interface MockCall {
  operationName: string | null;
  variables: Record<string, unknown>;
  rawPayload: string;
}

function makeQueue<T>() {
  const items: T[] = [];
  const pending: Array<(r: IteratorResult<T>) => void> = [];
  return {
    push(item: T) {
      const resolve = pending.shift();
      resolve ? resolve({ value: item, done: false }) : items.push(item);
    },
    get iterable(): AsyncIterable<T> {
      return {
        [Symbol.asyncIterator]() {
          return {
            next(): Promise<IteratorResult<T>> {
              if (items.length > 0) return Promise.resolve({ value: items.shift()!, done: false });
              return new Promise(r => pending.push(r));
            },
          };
        },
      };
    },
  };
}

export class MockAdapter {
  private readonly _responses = new Map<string, string>();
  private readonly _calls: MockCall[] = [];
  simulateFailure = false;

  register(operationName: string, jsonResponse: string): void {
    this._responses.set(operationName, jsonResponse);
  }

  registerData(operationName: string, data: Record<string, unknown>): void {
    this.register(operationName, JSON.stringify({ data }));
  }

  registerErrors(operationName: string, errors: unknown[]): void {
    this.register(operationName, JSON.stringify({ data: null, errors }));
  }

  get calls(): readonly MockCall[] { return this._calls; }
  get lastCall(): MockCall | null { return this._calls.at(-1) ?? null; }
  get callCount(): number { return this._calls.length; }

  reset(): void {
    this._calls.length = 0;
    this._responses.clear();
    this.simulateFailure = false;
  }

  readonly call = async (payload: string): Promise<string> => {
    if (this.simulateFailure) throw new Error('Simulated network failure');

    const decoded = JSON.parse(payload) as { query?: string; variables?: Record<string, unknown> };
    const opName = MockAdapter._extractOperationName(decoded.query ?? '');

    this._calls.push({ operationName: opName, variables: decoded.variables ?? {}, rawPayload: payload });

    const response = opName != null ? this._responses.get(opName) : undefined;
    if (response != null) return response;

    throw new Error(
      `MockAdapter: no response registered for "${opName ?? 'unknown'}".\n` +
      `Registered: [${[...this._responses.keys()].join(', ')}]\n` +
      `Payload: ${payload}`,
    );
  };

  private static _extractOperationName(query: string): string | null {
    return /(?:query|mutation|subscription)\s+(\w+)/.exec(query.trim())?.[1] ?? null;
  }
}

export class MockWsAdapter implements GraphLinkWsAdapter {
  private readonly _messages = makeQueue<string>();
  private readonly _reconnects = makeQueue<void>();
  private _subscriptionId: string | null = null;

  /**
   * Push a typed data payload as a graphql-ws `next` message.
   * Requires the subscription handshake to be complete first — call
   * `await wsAdapter.waitForSubscription()` before using this.
   */
  pushData(data: Record<string, unknown>): void {
    if (this._subscriptionId == null) throw new Error('MockWsAdapter: no active subscription id');
    this._messages.push(JSON.stringify({ type: 'next', id: this._subscriptionId, payload: { data } }));
  }

  /** Poll until the subscribe message has been received (handshake + subscribe done). */
  async waitForSubscription(timeoutMs = 1000): Promise<void> {
    const deadline = Date.now() + timeoutMs;
    while (this._subscriptionId == null) {
      if (Date.now() > deadline) throw new Error('Timeout waiting for subscription handshake');
      await new Promise(r => setTimeout(r, 5));
    }
  }

  /** Inject a raw message as if it arrived from the server. */
  push(message: string): void { this._messages.push(message); }

  /** Simulate a reconnect event. */
  reconnect(): void { this._reconnects.push(undefined); }

  async connect(): Promise<void> {}

  async sendMessage(message: string): Promise<void> {
    const parsed = JSON.parse(message) as { type: string; id?: string };
    if (parsed.type === 'connection_init') {
      // Respond with connection_ack after current call stack unwinds so the
      // ack listener is registered before the message is dispatched.
      setTimeout(() => this._messages.push(JSON.stringify({ type: 'connection_ack' })), 0);
    } else if (parsed.type === 'subscribe') {
      this._subscriptionId = parsed.id ?? null;
    }
  }

  async close(): Promise<void> {}
  async connectionInitPayload(): Promise<Record<string, unknown> | null> { return null; }

  get onMessageStream(): AsyncIterable<string> { return this._messages.iterable; }
  get onReconnect(): AsyncIterable<void> { return this._reconnects.iterable; }
}

File: integration_tests/dart_client_tests_real/test/mock_adapter.dart
import 'dart:async';
import 'dart:convert';

import 'package:graphlink_client_integration_tests/generated/client/graph_link_uploads.dart'
    show UploadProgressCallback;
import 'package:graphlink_client_integration_tests/generated/graphlink.dart';

/// One recorded outgoing GraphQL request.
class GraphLinkCall {
  final String? operationName;
  final Map<String, dynamic> variables;
  final String rawPayload;

  const GraphLinkCall({
    required this.operationName,
    required this.variables,
    required this.rawPayload,
  });
}

/// Drop-in adapter for [GraphLinkClient] that returns pre-registered responses
/// and records every outgoing request for assertion.
///
/// Usage:
///   final adapter = MockAdapter();
///   adapter.registerData('getUser', {'getUser': userJson});
///   final client = GraphLinkClient(adapter: adapter.call, wsAdapter: MockWebSocketAdapter());
///   // ... call client methods ...
///   expect(adapter.callCount, equals(1));
///   expect(adapter.lastCall!.variables['id'], equals('user-1'));
class MockAdapter {
  final _responses = <String, String>{};
  final _calls = <GraphLinkCall>[];

  /// Set to true to make every call throw — used for staleIfOffline tests.
  bool simulateFailure = false;

  /// Register a raw JSON string response for [operationName].
  void register(String operationName, String jsonResponse) {
    _responses[operationName] = jsonResponse;
  }

  /// Wraps [data] as `{"data": data}` and registers it.
  void registerData(String operationName, Map<String, dynamic> data) =>
      register(operationName, jsonEncode({'data': data}));

  /// Registers a GraphQL error response (data is null).
  void registerErrors(
    String operationName,
    List<Map<String, dynamic>> errors,
  ) =>
      register(operationName, jsonEncode({'data': null, 'errors': errors}));

  List<GraphLinkCall> get calls => List.unmodifiable(_calls);
  GraphLinkCall? get lastCall => _calls.isEmpty ? null : _calls.last;
  int get callCount => _calls.length;

  void reset() {
    _calls.clear();
    _responses.clear();
    simulateFailure = false;
  }

  Future<String> call(String payload) async {
    if (simulateFailure) throw Exception('Simulated network failure');

    final decoded = jsonDecode(payload) as Map<String, dynamic>;
    final query = decoded['query'] as String? ?? '';
    final variables =
        (decoded['variables'] as Map<String, dynamic>?) ?? const {};
    final opName = _extractOperationName(query);

    _calls.add(GraphLinkCall(
      operationName: opName,
      variables: variables,
      rawPayload: payload,
    ));

    if (opName != null && _responses.containsKey(opName)) {
      return _responses[opName]!;
    }

    throw StateError(
      'MockAdapter: no response registered for "${opName ?? "unknown"}".\n'
      'Registered operations: ${_responses.keys.toList()}\n'
      'Payload: $payload',
    );
  }

  static String? _extractOperationName(String query) =>
      _parseOperationName(query);
}

String? _parseOperationName(String query) {
  final match = RegExp(r'(?:query|mutation|subscription)\s+(\w+)')
      .firstMatch(query.trim());
  return match?.group(1);
}

/// Multipart adapter for upload mutation tests.
///
/// Intercepts `GLMultipartAdapter` calls, returns pre-registered JSON
/// responses, and records the raw `parts` map for assertion.
///
/// Usage:
///   final multipart = MockMultipartAdapter();
///   multipart.registerData('uploadAvatar', {'uploadAvatar': userJson});
///   final client = GraphLinkClient(
///     adapter: adapter.call,
///     uploadAdapter: multipart.call,
///     wsAdapter: MockWebSocketAdapter(),
///   );
class MockMultipartAdapter {
  final _responses = <String, String>{};
  final _calls = <Map<String, Object>>[];

  /// Register a raw JSON string response for [operationName].
  void register(String operationName, String jsonResponse) {
    _responses[operationName] = jsonResponse;
  }

  /// Wraps [data] as `{"data": data}` and registers it.
  void registerData(String operationName, Map<String, dynamic> data) =>
      register(operationName, jsonEncode({'data': data}));

  List<Map<String, Object>> get calls => List.unmodifiable(_calls);
  Map<String, Object>? get lastCall => _calls.isEmpty ? null : _calls.last;
  int get callCount => _calls.length;

  void reset() {
    _calls.clear();
    _responses.clear();
  }

  Future<String> call(
      Map<String, Object> parts, UploadProgressCallback? onProgress) async {
    _calls.add(Map.unmodifiable(parts));
    final decoded =
        jsonDecode(parts['operations'] as String) as Map<String, dynamic>;
    final opName = _parseOperationName(decoded['query'] as String? ?? '');

    if (opName != null && _responses.containsKey(opName)) {
      return _responses[opName]!;
    }

    throw StateError(
      'MockMultipartAdapter: no response registered for "${opName ?? "unknown"}".\n'
      'Registered operations: ${_responses.keys.toList()}',
    );
  }
}

/// WebSocket adapter for tests. Automatically handles the graphql-ws
/// handshake so tests only have to push data events and assert results.
class MockWebSocketAdapter extends GraphLinkWebSocketAdapter {
  final _controller = StreamController<String>.broadcast();
  final _reconnectController = StreamController<void>.broadcast();
  final _sent = <Map<String, dynamic>>[];

  String? _subscriptionId;
  final _ready = Completer<void>();

  /// All raw messages sent by the client (connection_init, subscribe, …).
  List<Map<String, dynamic>> get sentMessages => List.unmodifiable(_sent);

  /// Completes once the client has sent the graphql-ws "subscribe" message,
  /// i.e. the subscription stream is fully wired and ready to receive events.
  Future<void> get subscriptionReady => _ready.future;

  /// Push a graphql-ws "next" event for the active subscription.
  void pushEvent(Map<String, dynamic> data) {
    _controller.add(jsonEncode({
      'id': _subscriptionId,
      'type': 'next',
      'payload': {'data': data},
    }));
  }

  /// Push a "complete" event to close the active subscription stream.
  void pushComplete() {
    _controller.add(jsonEncode({'id': _subscriptionId, 'type': 'complete'}));
  }

  /// Push a raw JSON string as if it arrived from the server.
  void push(String message) => _controller.add(message);

  /// Simulate a reconnect event.
  void reconnect() => _reconnectController.add(null);

  @override
  Future<void> connect() async {}

  @override
  Stream<String> get onMessageStream => _controller.stream;

  @override
  Stream<void> get onReconnect => _reconnectController.stream;

  @override
  Future<void> sendMessage(String message) async {
    final decoded = jsonDecode(message) as Map<String, dynamic>;
    _sent.add(decoded);
    final type = decoded['type'] as String?;
    if (type == 'connection_init') {
      // Defer the ack by one event-loop turn so that _performHandshake's
      // firstWhere listener is already attached when the ack arrives on the
      // broadcast stream.
      Future.delayed(Duration.zero, () {
        _controller.add(jsonEncode({'type': 'connection_ack'}));
      });
    } else if (type == 'subscribe') {
      _subscriptionId = decoded['id'] as String?;
      if (!_ready.isCompleted) _ready.complete();
    }
  }

  @override
  Future<void> close() async {
    await _controller.close();
    await _reconnectController.close();
  }
}
